src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -414,5 +432,95 @@ private static void CheckForCompletedTransfers(SafeCurlMultiHandle multiHandle,
                 }
             }
         }
+
+        private async static void PollFunction(SafeCurlMultiHandle multiHandle)
+        {
+            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+            while (true)
+            {
+                if (cancellationTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                int count = 0;
+                lock (multiHandle)
+                {
+                    multiHandle.CancellationTokenSource = cancellationTokenSource;
+                    if (null != multiHandle.FdSet)
+                    {
+                        count = multiHandle.FdSet.Count;
+                    }
+                }
+                if (count == 0)
+                {
+                    // This can happen rarely when the socket callback is not
+                    // yet invoked for the first request
+                    await Task.Delay(100).ConfigureAwait(false);
+                    continue;
+                }
+
+                int result = -1;
+                unsafe
+                {
+                    Interop.libc.pollfd* fds = stackalloc Interop.libc.pollfd[count];
+                    lock (multiHandle)
+                    {
+                        if (count != multiHandle.FdSet.Count)
+                        {
+                            continue;
+                        }
+                        int i = 0;
+                        foreach (int fd in multiHandle.FdSet)
+                        {
+                            fds[i].fd = fd;
+                            fds[i].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                            i++;
+                        }
+                    }
+
+                    var retVal = Interop.libc.poll(fds, (uint)count, -1);
```

Comment:
Thanks, @stephentoub . This is a good idea. I have made this change and am sending for some early feedback.